File: CPP/DFS/733. Flood Fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int m = image.size(), n = image[0].size();
        dfs(image, sr, sc, m, n, image[sr][sc], color);
        return image;
    }
    void dfs(vector<vector<int>>& image, int r, int c, int m, int n,int target, int newColor){
        if (r < 0 || c < 0 || r >= m || c >= n || image[r][c] != target || image[r][c] == newColor) return;
        image[r][c] = newColor;
        dfs(image, r-1, c, m, n, target, newColor);
        dfs(image, r+1, c, m, n, target, newColor);
        dfs(image, r, c-1, m, n, target, newColor);
        dfs(image, r, c+1, m, n, target, newColor);
    }
};

File: Python/DFS/733. Flood Fill.py
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        start = image[sr][sc]
        #dfs line
        def dfs(i, j, memo):
            if i < 0 or j < 0 or i >= len(image) or j >= len(image[0]) or image[i][j] != start:
                return
            
            if (i,j) in memo:
                return
            else:
                if image[i][j] == start:
                    image[i][j] = color
                    
            memo[(i,j)] = 1
            dfs(i + 1, j, memo)
            dfs(i - 1, j, memo)
            dfs(i, j + 1, memo)
            dfs(i, j - 1, memo)
        
        if start != color:
            dfs(sr, sc, {})
        
        return image
        

        
        
if __name__ == '__main__':
    image = [[1,1,1],[1,1,0],[1,0,1]]
    sr = 1
    sc = 1
    color = 2
    sol = Solution()
    print(sol.floodFill(image, sr, sc, color))